File: binance.py

import requests
import zipfile
import os
import shutil
import pandas as pd
import mysql.connector
from multiprocessing import Pool, cpu_count
from urllib.parse import urlparse
from datetime import datetime, timedelta
from dotenv import load_dotenv

load_dotenv()

# Database configuration
db_config = {
    'host': os.getenv("DB_HOST"),
    'user': os.getenv("DB_USER"),
    'password': os.getenv("DB_PASSWORD"),
    'database': 'trading'
}

# List of URLs
file_urls = [
    'https://data.binance.vision/data/spot/daily/klines/PEPEUSDC/1m/PEPEUSDC-1m-',
    # Add more URLs here
]

def generate_urls(base_url, start_date, end_date=None):
    start = datetime.strptime(start_date, '%Y-%m-%d')
    end = datetime.strptime(end_date, '%Y-%m-%d') if end_date else datetime.now() - timedelta(days=1)
    
    urls = []
    current_date = start

    while current_date <= end:
        date_str = current_date.strftime('%Y-%m-%d')
        urls.append(f"{base_url}{date_str}.zip")
        current_date += timedelta(days=1)

    return urls

def download_and_unzip(url):
    # Parse the file name from the URL
    file_name = os.path.basename(urlparse(url).path)
    zip_path = os.path.join('downloads', file_name)
    extract_dir = zip_path.replace('.zip', '')
    csv_file = os.path.join(extract_dir, file_name.replace('.zip', '.csv'))
    
    if os.path.exists(csv_file):
        print(f"Skipping {url}: Already downloaded")
        return csv_file
    

    # Download the file
    print(f"Downloading {url}")
    response = requests.get(url, stream=True)
    if response.raise_for_status():
        print(f"Failed to download {url}")
        return
    with open(zip_path, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)

    # Unzip the file
    print(f"Unzipping {zip_path}")
    with zipfile.ZipFile(zip_path, 'r') as zip_ref:
        zip_ref.extractall(extract_dir)

    # Remove the zip file after extraction
    os.remove(zip_path)

    # Return the path to the CSV file
    return csv_file

def load_csv_to_database(csv_file, exchange, token):
    print(f"Loading data from {csv_file} to database")
    connection = mysql.connector.connect(**db_config)
    cursor = connection.cursor()

    # Read CSV into pandas for processing
    data = pd.read_csv(csv_file)
    for _, row in data.iterrows():
        sql = """INSERT INTO KLINES_SPOT (exchange, token, open_time, open, high, low, close, volume, close_time,
                 quote_asset_volume, number_of_trades, taker_buy_base_asset_volume, taker_buy_quote_asset_volume)
                 VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
                 ON DUPLICATE KEY UPDATE open=open"""  # Adjust the update clause as needed
        cursor.execute(sql, (exchange, token, *row[:-1]))

    connection.commit()
    cursor.close()
    connection.close()
    print(f"Data loaded from {csv_file}")

def process_file(url):
    try:
        csv_file = download_and_unzip(url)
        load_csv_to_database(csv_file, 'Binance', 'PEPEUSDC')
    except Exception as e:
        print(f"Failed to process {url}: {e}")

def main():
    # Generate URLs for the specified date range
    start_date = '2024-09-01'
    urls = generate_urls(file_urls[0], start_date)

    os.makedirs('downloads', exist_ok=True)

    # Use multiprocessing Pool to process files in parallel
    with Pool(cpu_count()) as pool:
        pool.map(process_file, urls)

if __name__ == '__main__':
    main()

File: binance.js
const unzipper = require('unzipper');
const mysql = require('mysql2/promise');
const fs = require('fs');
const path = require('path');
const readline = require('readline');
const request = require('request');

const datasets = [
    { 
        path: "https://data.binance.vision/data/spot/daily/klines/PEPEUSDC/1m/", 
        filenamePrefix: "PEPEUSDC-1m",
        exchange: "BINANCE",
        token: "PEPE",
        type: "SPOT",
    },
    // {   
    //     path: "https://data.binance.vision/data/futures/um/daily/klines/1000PEPEUSDC/1m/", 
    //     filenamePrefix: "1000PEPEUSDC-1m",
    //     exchange: "BINANCE",
    //     token: "PEPE",
    //     type: "PERP"
    // },
    // { 
    //     path: "https://data.binance.vision/data/futures/um/daily/metrics/1000PEPEUSDC/", 
    //     filenamePrefix: "1000PEPEUSDC-metrics",
    //     exchange: "BINANCE",
    //     token: "PEPE",
    //     type: "METRICS"
    // }
];

const startYear = 2024;
const startMonth = 9; // September (1-based index)

// MySQL configuration
const dbConfig = {
    host: process.env.DB_HOST,
    user: process.env.DB_USER,
    password: process.env.DB_PASSWORD,
    database: "trading",
};

// Global MySQL connection
let connection;

// Initialize the database connection
async function initDbConnection() {
    connection = await mysql.createConnection(dbConfig);
    console.log('Database connection established.');
}

function generateUrlsFromDateRange(datasets) {
    const urls = [];
    const today = new Date();
    let year = startYear;
    let month = startMonth;

    while (year < today.getFullYear() || (year === today.getFullYear() && month <= today.getMonth() + 1)) {
        const monthStr = month.toString().padStart(2, '0');
        const daysInMonth = new Date(year, month, 0).getDate();

        // Generate daily URLs for each dataset
        for (const dataset of datasets) {
            const tableName = dataset.type === 'METRICS' ? 'METRICS' : 'KLINES_'+dataset.type;
            for (let day = 1; day <= daysInMonth; day++) {
                const dayStr = day.toString().padStart(2, '0');
                const dateStr = `${year}-${monthStr}-${dayStr}`;
                const filename = `${dataset.filenamePrefix}-${dateStr}.zip`;
                // check if the file exists (change zip to csv)
                if (fs.existsSync(filename.replace('.zip', '.csv'))) {
                    console.log(`File ${filename} already exists`);
                    continue;
                }
                urls.push({
                    url: `${dataset.path}${filename}`,
                    tableName: tableName,
                    exchange: dataset.exchange,
                    token: dataset.token,
                });
            }
        }

        month++;
        if (month > 12) {
            month = 1;
            year++;
        }
    }
    return urls;
}

function downloadZip(url) {
    console.log(`Starting download from ${url}`);

    const output = path.join(process.cwd(), url.split('/').pop());
    console.log(`Output path: ${output}`);

    return new Promise((resolve, reject) => {
        request({ url: url, encoding: null }, (err, resp, body) => {
            if (err) {
                return reject(err);
            }

            fs.writeFile(output, body, (err) => {
                if (err) {
                    return reject(err);
                }
                console.log("File downloaded and written to disk.");
                resolve(output);
            });
        });
    });
}

function unzipFile(filePath) {
    // Determine the expected CSV file name by replacing .zip with .csv
    const csvFileName = path.basename(filePath, '.zip') + '.csv';
    const extractDir = path.join(process.cwd(), path.basename(filePath, '.zip'));

    return new Promise((resolve, reject) => {
        fs.createReadStream(filePath)
            .pipe(unzipper.Extract({ path: "./" }))
            .on("close", () => {
                console.log("Files unzipped successfully.");

                // Delete the zip file after unzipping
                fs.unlinkSync(filePath);

                // Resolve with the CSV file name (not the path)
                resolve(csvFileName);
            })
            .on("error", (err) => reject(err));
    });
}

async function loadCsvToDatabase(csvFilePath, tableName, exchange, token) {
    console.log(`Loading data from ${csvFilePath}`);

    const rl = readline.createInterface({
        input: fs.createReadStream(csvFilePath),
        crlfDelay: Infinity
    });

    for await (const line of rl) {
        let data = line.split(',');  // Adjust based on CSV column structure
        data.unshift(token);
        data.unshift(exchange);
        data.pop();
        const query = `INSERT INTO ${tableName} (
                        exchange, token, open_time, open, high, low, close, volume, close_time,
                        quote_asset_volume, number_of_trades, taker_buy_base_asset_volume, 
                        taker_buy_quote_asset_volume
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
        try {
            await connection.query(query, data);
        } catch {
            console.log(`Error loading data from ${csvFilePath}`);
            break;
        }
    }
}

async function processData(url, tableName, exchange, token) {
    try {
        const zipFilePath = await downloadZip(url);
        const extractPath = await unzipFile(zipFilePath);

        const csvFilePath = path.join("./", extractPath);
        await loadCsvToDatabase(csvFilePath, tableName, exchange, token);
        console.log(`Data from ${csvFile} loaded successfully.`);

    } catch (error) {
        console.error(`Error processing data from ${url}: ${error.message}`);
    }
}

async function main() {
    try {
        await initDbConnection();

        // Generate URLs for each day from September onwards
        const urls = generateUrlsFromDateRange(datasets);

        // Process each generated URL
        for (const { url, tableName, exchange, token } of urls) {
            await processData(url, tableName, exchange, token);
        }

    } catch (error) {
        console.error('Error:', error.message);
    } finally {
        if (connection) {
            await connection.end();
            console.log('Database connection closed.');
        }
    }
}

// Run the main function
main().catch(console.error);